Mathlib/Topology/Algebra/Module/Action.lean

```diff
@@ -0,0 +1,253 @@
+     `continuousSMul_sInf`.
+  -/
+  continuousSMul_sInf <| fun _ _ ↦ by simp_all only [Set.mem_setOf_eq]
+
+/-- Addition `+ : A × A → A` is continuous if `R` is a topological
+ring, and `A` is an `R` module with the module topology. -/
+theorem ModuleTopology.continuousAdd : @ContinuousAdd A (moduleTopology R A) _ :=
+  continuousAdd_sInf <| fun _ _ ↦ by simp_all only [Set.mem_setOf_eq]
+
+instance IsModuleTopology.toContinuousSMul [TopologicalSpace A] [IsModuleTopology R A] :
+    ContinuousSMul R A := eq_moduleTopology R A ▸ ModuleTopology.continuousSMul R A
+
+-- this can't be an instance because typclass inference can't be expected to find `R`.
+theorem IsModuleTopology.toContinuousAdd [TopologicalSpace A] [IsModuleTopology R A] :
+    ContinuousAdd A := eq_moduleTopology R A ▸ ModuleTopology.continuousAdd R A
+
+/-- The module topology is `≤` any topology making `A` into a topological module. -/
+theorem moduleTopology_le [τA : TopologicalSpace A] [ContinuousSMul R A] [ContinuousAdd A] :
+    moduleTopology R A ≤ τA := sInf_le ⟨inferInstance, inferInstance⟩
+
+end basics
+
+namespace IsModuleTopology
+
+section subsingleton
+
+instance instSubsingleton (R : Type*) [TopologicalSpace R] (A : Type*) [Add A] [SMul R A]
+    [Subsingleton A] [TopologicalSpace A] : IsModuleTopology R A where
+  eq_moduleTopology' := by
+    ext U
+    simp only [isOpen_discrete]
+
+end subsingleton
+
+section iso
+
+variable {R : Type*} [τR : TopologicalSpace R] [Semiring R]
+variable {A : Type*} [AddCommMonoid A] [Module R A] [τA : TopologicalSpace A] [IsModuleTopology R A]
+variable {B : Type*} [AddCommMonoid B] [Module R B] [τB : TopologicalSpace B]
+
+/-- If `A` and `B` are homeomorphic via a homeomorphism which is also `R`-linear, and if
+`A` has the module topology, then so does `B`. -/
+theorem iso (e : A ≃L[R] B) : IsModuleTopology R B where
+  eq_moduleTopology' := by
+    -- get these in before I start putting new topologies on A and B and have to use `@`
+    let g : A →ₗ[R] B := e
+    let g' : B →ₗ[R] A := e.symm
+    let h : A →+ B := e
+    let h' : B →+ A := e.symm
+    simp_rw [e.toHomeomorph.symm.inducing.1, eq_moduleTopology R A, moduleTopology, induced_sInf]
+    apply congr_arg
+    ext τ
+    rw [Set.mem_image]
+    constructor
+    · rintro ⟨σ, ⟨hσ1, hσ2⟩, rfl⟩
+      exact ⟨continuousSMul_induced g', continuousAdd_induced h'⟩
+    · rintro ⟨h1, h2⟩
+      use τ.induced e
+      rw [induced_compose]
+      refine ⟨⟨continuousSMul_induced g, continuousAdd_induced h⟩, ?_⟩
+      nth_rw 2 [← induced_id (t := τ)]
+      simp
+
+end iso
+
+section self
+
+variable (R : Type*) [Semiring R] [τR : TopologicalSpace R] [TopologicalSemiring R]
+
+/-!
+We now fix once and for all a topological semiring `R`.
+
+We first prove that the module topology on `R` considered as a module over itself,
+is `R`'s topology.
+-/
+
+/-- The topology on a topological semiring `R` agrees with the module topology when considering
+`R` as an `R`-module. -/
```

Comment:
```suggestion
`R` as an `R`-module via `Semiring.toModule`. -/
```
and similar for the instance below.

Reply:
For a mathematician reading that docstring, they will think nothing at all about the claim that a ring is a module over itself because it's the canonical example of a module and the first one that everyone sees, but if I start *explaining* why a ring is a module over itself by throwing around mathlibisms then their first thought may be "what the heck is `Semiring.toModule` and what is happening here? Is R is a module over itself in some weird way that the computer scientists have thought of"?